### PR TITLE
update elastic scaling guide

### PR DESCRIPTION
**Pull request by alindima**
Resolves https://github.com/paritytech/polkadot-sdk/issues/5050

Updates the elastic scaling guide, taking into consideration:
- the completed implementation of [RFC-103](https://github.com/polkadot-fellows/RFCs/pull/103), which enables an untrusted collator set for elastic scaling. Adds the necessary instructions for configuring the collator so that it can leverage this implementation
- general updates for bits that became out of date

This PR should not be merged until:
1. the `CandidateReceiptV2` node feature bit is enabled on all networks
2. the functionality hidden under the `experimental-ump-signals` feature of the parachain-system pallet is turned on by default (which can only be done after 1)

TODO:
- [ ] update after https://github.com/paritytech/polkadot-sdk/pull/6939 is merged
- [x] add info about the possibility of needing to modify authoring duration if using more than 3 cores

### CODE REVIEW

File: docs/sdk/src/guides/enable_elastic_scaling_mvp.rs [hunk 1]
```diff
@@ -89,17 +82,63 @@
...
+//! on. It does so by implementing the `SelectCore` trait.
+#![doc = docify::embed!("../../cumulus/pallets/parachain-system/src/lib.rs", SelectCore)]
+//!
+//! For the vast majority of use cases though, you will not need to implement a custom core
+//! selector. There are two core selection policies to choose from (without implementing your own)
+//! `DefaultCoreSelector` and `LookaheadCoreSelector`.
+//!
+//! - The `DefaultCoreSelector` implements a round-robin selection on the cores that can be
+//! occupied by the parachain at the very next relay parent. This is the equivalent to what all
+//! parachains on production networks have been using so far.
```
Comment (eskimor) [inline docs/sdk/src/guides/enable_elastic_scaling_mvp.rs:111]: Hmm. Shall we rename this as part of this PR? It seems like LookaheadCoreSelector should be the "default" as we expect any new parachain to use asynchronous backing?

File: docs/sdk/src/guides/enable_elastic_scaling_mvp.rs [hunk 2]
```diff
@@ -108,8 +147,10 @@
 //! produce per relay chain block (in direct correlation with the number of acquired cores). This
 //! should be either 1 (no scaling), 2 or 3. This is called the parachain velocity.
 //!
-//! If you configure a velocity which is different from the number of assigned cores, the measured
-//! velocity in practice will be the minimum of these two.
+//! <div class="warning">If you configure a velocity which is different from the number of assigned
+//! cores, the measured velocity in practice will be the minimum of these two. However, be mindful
+//! that if the velocity is higher than the number of assigned cores, it's possible that
+//! <a href="https://github.com/paritytech/polkadot-sdk/issues/6667"> only a subset of the collator set will be authoring blocks.</a></div>
```
Comment (eskimor) [inline docs/sdk/src/guides/enable_elastic_scaling_mvp.rs:153]: The question is why do we need to configure a velocity at all, seems redundant.

Comment (eskimor) [reply docs/sdk/src/guides/enable_elastic_scaling_mvp.rs:153]: Once the slot based collator can produce multiple blocks per slot we should also add that we recommend slot durations of at least 6s, preferably even 12. (better censorship resistance)

File: docs/sdk/src/guides/enable_elastic_scaling.rs [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+//!     `task_manager.spawn_handle()` and pass in the `block_import_handle` param.
+//!     - Replace the `aura::run` with the `slot_based::run` call and remove the explicit task
+//!       spawn:
+#![doc = docify::embed!("../../cumulus/polkadot-omni-node/lib/src/nodes/aura.rs", launch_slot_based_collator)]
+//!
+//! 3. In `start_parachain_node()`, destructure `slot_based_handle` from `params.other`. Remove the
+//!    `overseer_handle` and `relay_chain_slot_duration` params passed to `start_consensus` and pass
+//!    in the `slot_based_handle`.
+//!
+//! ### Phase 2 - Configure core selection policy in the parachain runtime
```
Comment (alindima) [inline docs/sdk/src/guides/enable_elastic_scaling.rs:150]: Phase 2 assumes candidate receipt v2 feature bit was enabled.
This phase will change after the feature bit is enabled on all networks and a form of https://github.com/paritytech/polkadot-sdk/pull/6939 is merged

File: templates/parachain/node/src/service.rs [hunk 4]
```diff
@@ -15,7 +15,9 @@ use polkadot_sdk::*;
 use cumulus_client_cli::CollatorOptions;
 use cumulus_client_collator::service::CollatorService;
 #[docify::export(lookahead_collator)]
-use cumulus_client_consensus_aura::collators::lookahead::{self as aura, Params as AuraParams};
+use cumulus_client_consensus_aura::collators::slot_based::{
```
Comment (alindima) [inline templates/parachain/node/src/service.rs:18]: Changes in this file will be rolled back before merge, but currently showcase what a parachain team using the template would need to do on the node-side to use elastic scaling

File: docs/sdk/src/guides/enable_elastic_scaling.rs [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+//!
```
Comment (sandreim) [inline docs/sdk/src/guides/enable_elastic_scaling.rs:154]: Let's clearly mention that RFC103 is mandatory for launching in production.

File: docs/sdk/src/guides/enable_elastic_scaling.rs [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+//! impl cumulus_primitives_core::GetCoreSelectorApi<Block> for Runtime {
+//! 		fn core_selector() -> (cumulus_primitives_core::CoreSelector, cumulus_primitives_core::ClaimQueueOffset) {
+//! 			ParachainSystem::core_selector()
+//! 		}
+//! }
+//! ```
+//!
+//! ### Phase 3 - Configure maximum scaling factor in the runtime
+//!
+//! First of all, you need to decide the upper limit to how many parachain blocks you need to
```
Comment (sandreim) [inline docs/sdk/src/guides/enable_elastic_scaling.rs:199]: Actually the thinking is the other way around - what is the minimum target block time? It is then no longer needed to configure any other parameters manually as you can compute them from this value.

File: docs/sdk/src/guides/enable_elastic_scaling.rs [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+//!
+//! [Elastic scaling](https://polkadot.network/blog/elastic-scaling-streamling-growth-on-polkadot)
+//! is a feature that enables parachains to seamlessly scale up/down the number of used cores.
+//! This can be used to increase the compute or storage throughput of a parachain or
+//! to lower the latency between a transaction being submitted and it getting built in a parachain
+//! block.
+//!
+//! ## Current constraints
+//!
+//! Elastic scaling is still considered experimental software, so stability is not guaranteed.
```
Comment (sandreim) [inline docs/sdk/src/guides/enable_elastic_scaling.rs:17]: After launching on Polkadot this is not true.

File: docs/sdk/src/guides/enable_elastic_scaling.rs [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+//! If you encounter any problems,
+//! [please open an issue](https://github.com/paritytech/polkadot-sdk/issues).
+//! Below are described the constraints of the implementation:
+//!
+//! 1. **Bounded compute throughput**. Each parachain block gets at most 2 seconds of execution on
+//!    the relay chain. Therefore, assuming the full 2 seconds are used, a parachain can only
+//!    utilise at most 3 cores in a relay chain slot of 6 seconds. If the full execution time is not
+//!    being used, or if collators are able to author blocks faster than the validators are able to
+//!    run them, higher core counts can be achieved.
+//! 2. **Single collator requirement for consistently scaling beyond a core at full authorship
```
Comment (sandreim) [inline docs/sdk/src/guides/enable_elastic_scaling.rs:27]: Let's clearly mention single collator should only be used for testing as it means maximum centralisation. For production, multiple collators need to be used and this is what projects have in practice. Obviously the amount of compute you can use from each core depends on the latency between collators and the performance of the collators.

File: docs/sdk/src/guides/enable_elastic_scaling.rs [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+//!    adds additional latency to the block production pipeline. Assuming block execution takes
+//!    about the same as authorship, the additional overhead is equal the duration of the authorship
+//!    plus the block announcement. Each collator must first import the previous block before
+//!    authoring a new one, so it is clear that the highest throughput can be achieved using a
+//!    single collator. Experiments show that the peak performance using more than one collator
+//!    (measured up to 10 collators) is utilising 2 cores with authorship time of 1.3 seconds per
+//!    block, which leaves 400ms for networking overhead. This would allow for 2.6 seconds of
+//!    execution, compared to the 2 seconds async backing enabled.
+//!    The development required for enabling maximum compute throughput for multiple collators is tracked by
+//!    [this issue](https://github.com/paritytech/polkadot-sdk/issues/5190).
```
Comment (sandreim) [inline docs/sdk/src/guides/enable_elastic_scaling.rs:38]: I think we can do much better in terms of structure here vs a large blob of text which is not that easy to read and focus important information.

File: docs/sdk/src/guides/enable_elastic_scaling.rs [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+//! 3. **Lack of out-of-the-box automated scaling.** For true elasticity, the parachain must be able
+//!    to seamlessly acquire or sell coretime as the user demand grows and shrinks over time, in an
+//!    automated manner. This is
+//!    currently lacking - a parachain can only scale up or down by “manually” acquiring coretime.
+//!    This is not in the scope of the relay chain functionality. Parachains can already start
+//!    implementing such autoscaling, but we aim to provide a framework/examples for developing
+//!    autoscaling strategies.
+//!    Tracked by [this issue](https://github.com/paritytech/polkadot-sdk/issues/1487).
```
Comment (sandreim) [inline docs/sdk/src/guides/enable_elastic_scaling.rs:46]: We should do an update: Phase 1 seems done: https://docs.google.com/document/d/1v23ViA7VggRz3StMsWqoRozNgTxU_ylrao3tSM8_Cjc/edit?tab=t.0

File: docs/sdk/src/guides/enable_elastic_scaling.rs [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+//!
+//! Another hard limitation that is not envisioned to ever be lifted is that parachains which create
+//! forks will generally not be able to utilise the full number of cores they acquire.
```
Comment (sandreim) [inline docs/sdk/src/guides/enable_elastic_scaling.rs:49]: I would rephrase to just say something like "When the parachain forks it's throughput is decreased."

File: docs/sdk/src/guides/enable_elastic_scaling.rs [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+//!
+//! ## Using elastic scaling
+//!
+//! ### Prerequisites
+//!
+//! - Ensure Asynchronous Backing is enabled on the relay chain network and you have enabled it on
+//!   the parachain using [`crate::guides::async_backing_guide`].
```
Comment (sandreim) [inline docs/sdk/src/guides/enable_elastic_scaling.rs:56]: Can't we just remove this ? All networks have it.

File: docs/sdk/src/guides/enable_elastic_scaling.rs [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+//! - Ensure the `AsyncBackingParams.max_candidate_depth` value is configured to a value that is at
+//!   least double the maximum targeted parachain velocity. For example, if the parachain will build
+//!   at most 3 candidates per relay chain block, the `max_candidate_depth` should be at least 6.
+//! - Ensure enough coretime is assigned to the parachain. For maximum compute throughput the upper
+//!   bound is 3 cores.
```
Comment (sandreim) [inline docs/sdk/src/guides/enable_elastic_scaling.rs:61]: What about latency ? 500ms blocks ?

File: docs/sdk/src/guides/enable_elastic_scaling.rs [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+//! latest polkadot-sdk release! Simply pass the <code>--experimental-use-slot-based</code>
+//! ([`polkadot_omni_node_lib::cli::Cli::experimental_use_slot_based`]) parameter to the command
+//! line and jump to Phase 2.</div>
+//!
+//! ### Phase 1 - (For custom parachain node) Update Parachain Node
+//!
+//! This assumes you are using
+//! [the latest parachain template](https://github.com/paritytech/polkadot-sdk/tree/master/templates/parachain).
+//!
+//! This phase consists of plugging in the new slot-based collator.
```
Comment (sandreim) [inline docs/sdk/src/guides/enable_elastic_scaling.rs:76]: Let's mention somewhere that the slot based collator is a hard requirement, Elastic Scaling doesn't work without.

File: docs/sdk/src/guides/enable_elastic_scaling.rs [hunk 15]
```diff
@@ -0,0 +1,238 @@
...
+//!
+//! ## Quick introduction to elastic scaling
+//!
+//! [Elastic scaling](https://polkadot.network/blog/elastic-scaling-streamling-growth-on-polkadot)
+//! is a feature that enables parachains to seamlessly scale up/down the number of used cores.
+//! This can be used to increase the compute or storage throughput of a parachain or
+//! to lower the latency between a transaction being submitted and it getting built in a parachain
+//! block.
+//!
+//! ## Current constraints
```
Comment (sandreim) [inline docs/sdk/src/guides/enable_elastic_scaling.rs:15]: I would rename to performance characteristics.

File: docs/sdk/src/guides/enable_elastic_scaling.rs [hunk 16]
```diff
@@ -0,0 +1,238 @@
...
+//!
+//! ```ignore
+//! impl cumulus_primitives_core::GetCoreSelectorApi<Block> for Runtime {
+//! 		fn core_selector() -> (cumulus_primitives_core::CoreSelector, cumulus_primitives_core::ClaimQueueOffset) {
+//! 			ParachainSystem::core_selector()
+//! 		}
+//! }
+//! ```
+//!
+//! ### Phase 3 - Configure maximum scaling factor in the runtime
```
Comment (sandreim) [inline docs/sdk/src/guides/enable_elastic_scaling.rs:197]: can't we link some sample test parachain already configured and a zombienet test from CI suite ?

File: docs/sdk/src/guides/enable_elastic_scaling.rs (hunk 6 above)
Comment (alindima) [reply docs/sdk/src/guides/enable_elastic_scaling.rs:199]: you can also make all the calculations based on the velocity, which is what I describe here

File: docs/sdk/src/guides/enable_elastic_scaling.rs (hunk 7 above)
Comment (alindima) [reply docs/sdk/src/guides/enable_elastic_scaling.rs:17]: True, will update when that is the case

File: docs/sdk/src/guides/enable_elastic_scaling.rs (hunk 12 above)
Comment (alindima) [reply docs/sdk/src/guides/enable_elastic_scaling.rs:56]: Definitely not all parachains have it

File: docs/sdk/src/guides/enable_elastic_scaling.rs (hunk 15 above)
Comment (alindima) [reply docs/sdk/src/guides/enable_elastic_scaling.rs:15]: I don't think it fits. Automated core procurement is not a performance consideration

File: docs/sdk/src/guides/enable_elastic_scaling.rs (hunk 9 above)
Comment (alindima) [reply docs/sdk/src/guides/enable_elastic_scaling.rs]: I rewrote this section. let me know how it looks

File: docs/sdk/src/guides/enable_elastic_scaling.rs (hunk 10 above)
Comment (alindima) [reply docs/sdk/src/guides/enable_elastic_scaling.rs:48]: I mentioned the regionx in-progress implementation

File: docs/sdk/src/guides/enable_elastic_scaling.rs (hunk 14 above)
Comment (alindima) [reply docs/sdk/src/guides/enable_elastic_scaling.rs]: I explicitly mentioned that slot based collator is mandatory

File: docs/sdk/src/guides/enable_elastic_scaling.rs (hunk 15 above)
Comment (alindima) [reply docs/sdk/src/guides/enable_elastic_scaling.rs]: Renamed to `Performance characteristics and constraints`

File: docs/sdk/src/guides/enable_elastic_scaling.rs (hunk 16 above)
Comment (alindima) [reply docs/sdk/src/guides/enable_elastic_scaling.rs:205]: added

File: docs/sdk/src/guides/enable_elastic_scaling.rs [hunk 17]
```diff
@@ -0,0 +1,249 @@
...
+//!    and resilience. Experiments show that the peak compute throughput using more than one collator
+//!    (measured up to 10 collators) is utilising 2 cores with authorship time of 1.3 seconds per block,
+//!    which leaves 400ms for networking overhead. This would allow for 2.6 seconds of execution, compared
+//!    to the 2 seconds async backing enabled. If block authoring duration is low and you attempt to
+//!    use elastic scaling for achieving low latency or increasing storage throughput, this is not a
+//!    problem. Developments required for streamlining block production are tracked by [this issue](https://github.com/paritytech/polkadot-sdk/issues/5190).
+//! 3. **Lack of out-of-the-box automated scaling.** For true elasticity, the parachain must be able
+//!    to seamlessly acquire or sell coretime as the user demand grows and shrinks over time, in an
+//!    automated manner. This is
+//!    currently lacking - a parachain can only scale up or down by “manually” acquiring coretime.
```
Comment (sandreim) [inline docs/sdk/src/guides/enable_elastic_scaling.rs:44]:
```suggestion
//!    currently lacking - a parachain can only scale up or down by implementing some off-chain solution for managing the core time resources.
```

File: docs/sdk/src/guides/enable_elastic_scaling.rs [hunk 18]
```diff
@@ -0,0 +1,249 @@
...
+//!    This is not in the scope of the relay chain functionality. Parachains can already start
+//!    implementing such autoscaling, but we aim to provide a framework/examples for developing
+//!    autoscaling strategies.
+//!    Tracked by [this issue](https://github.com/paritytech/polkadot-sdk/issues/1487).
+//!    An in-progress external implementation by RegionX can be found [here](https://github.com/RegionX-Labs/On-Demand).
+//!
+//! Another important constraint is that when a parachain forks, the number of blocks backed per
+//! relay chain block decreases.
```
Comment (sandreim) [inline docs/sdk/src/guides/enable_elastic_scaling.rs:52]: It is much easier for someone to understand if we mention clearly the impact first and the technical reason after.
```suggestion
//! Another important constraint is that when a parachain forks, the throughput decreases and latency increases because the number of blocks backed per
//! relay chain block decreases.
```

File: docs/sdk/src/guides/enable_elastic_scaling.rs (hunk 12 above)
Comment (sandreim) [reply docs/sdk/src/guides/enable_elastic_scaling.rs:63]: Indeed, what I am saying is to remove the part about the relay chain feature check.

File: docs/sdk/src/guides/enable_elastic_scaling.rs (hunk 6 above)
Comment (sandreim) [reply docs/sdk/src/guides/enable_elastic_scaling.rs:210]: I can see what is described here, but I want a better DX. 

As you've noticed recently, people didn't ask "how many parachains blocks can I produce per relay chain block ?", Instead they ask "How can I get 500ms blocks ?" because that is what their end users care about. The velocity of the parachain is largely an implementation detail.

With that being said, we can then remove all of the details about velocity and concerns around they need to compute all sorts of other constants.

File: docs/sdk/src/guides/enable_elastic_scaling.rs [hunk 19]
```diff
@@ -0,0 +1,249 @@
...
+//!
+//! <div class="warning">This guide assumes full familiarity with Asynchronous Backing and its
+//! terminology, as defined in <a href="https://wiki.polkadot.network/docs/maintain-guides-async-backing">the Polkadot Wiki</a>.
+//! Furthermore, the parachain should have already been upgraded according to the guide.</div>
+//!
+//! ## Quick introduction to elastic scaling
+//!
+//! [Elastic scaling](https://polkadot.network/blog/elastic-scaling-streamling-growth-on-polkadot)
+//! is a feature that enables parachains to seamlessly scale up/down the number of used cores.
+//! This can be used to increase the compute or storage throughput of a parachain or
```
Comment (sandreim) [inline docs/sdk/src/guides/enable_elastic_scaling.rs:11]:
```suggestion
//! This can be used to increase the available compute and bandwidth resources of a parachain or
```

File: docs/sdk/src/guides/enable_elastic_scaling.rs [hunk 20]
```diff
@@ -0,0 +1,249 @@
...
+//! to lower the latency between a transaction being submitted and it getting built in a parachain
+//! block.
```
Comment (sandreim) [inline docs/sdk/src/guides/enable_elastic_scaling.rs:13]:
```suggestion
//! to lower the transaction inclusion latency by decreasing block time.
```

File: docs/sdk/src/guides/enable_elastic_scaling.rs [hunk 21]
```diff
@@ -0,0 +1,249 @@
...
+//!
+//! Elastic scaling is still considered experimental software, so stability is not guaranteed.
+//! If you encounter any problems,
+//! [please open an issue](https://github.com/paritytech/polkadot-sdk/issues).
+//! Below are described the constraints of the implementation:
+//!
+//! 1. **Bounded compute throughput**. Each parachain block gets at most 2 seconds of execution on
+//!    the relay chain. Therefore, assuming the full 2 seconds are used, a parachain can only
+//!    utilise at most 3 cores in a relay chain slot of 6 seconds. If the full execution time is not
+//!    being used or if collators are able to author blocks faster than the validators are able to
```
Comment (sandreim) [inline docs/sdk/src/guides/enable_elastic_scaling.rs:25]: I think the correct way of saying this is: ` if all collators can author blocks faster than the reference hw` 

Relay chain validators are quite irrelevant as the validation happens in parallel and parachain blocks still have a cap of 2s of execution during backing. Weights on the parachain side actually ensure we don't hit backing timeout.

Comment (alindima) [reply docs/sdk/src/guides/enable_elastic_scaling.rs:25]: yeah, I was making the assumption that validators run reference hw

File: docs/sdk/src/guides/enable_elastic_scaling.rs [hunk 22]
```diff
@@ -0,0 +1,249 @@
...
+//! - Ensure the `AsyncBackingParams.max_candidate_depth` value is configured to a value that is at
+//!   least double the maximum targeted parachain velocity. For example, if the parachain will build
+//!   at most 3 candidates per relay chain block, the `max_candidate_depth` should be at least 6.
+//! - Ensure enough coretime is assigned to the parachain.
+//! - Ensure the `CandidateReceiptV2` node feature is enabled on the relay chain configuration (node
+//!   feature bit number 3).
+//!
+//! <div class="warning">Phase 1 is NOT needed if using the <code>polkadot-parachain</code> or
+//! <code>polkadot-omni-node</code> binary, or <code>polkadot-omni-node-lib</code> built from the
+//! latest polkadot-sdk release! Simply pass the <code>--experimental-use-slot-based</code>
```
Comment (skunert) [inline docs/sdk/src/guides/enable_elastic_scaling.rs:73]:
```suggestion
//! latest polkadot-sdk release! Simply pass the <code>--authoring slot-based</code>
```

File: docs/sdk/src/guides/enable_elastic_scaling.rs [hunk 23]
```diff
@@ -0,0 +1,249 @@
...
+//!
+//! 1. In `node/src/service.rs` import the slot based collator instead of the lookahead collator, as
+//!    well as the `SlotBasedBlockImport` and `SlotBasedBlockImportHandle`.
+#![doc = docify::embed!("../../cumulus/polkadot-omni-node/lib/src/nodes/aura.rs", slot_based_colator_import)]
+//!
+//! 2. Modify the `ParachainBlockImport` and `Service` type definitions:
+//! ```ignore
+//! type ParachainBlockImport = TParachainBlockImport<
+//! 	    Block,
+//! 	    SlotBasedBlockImport<Block, Arc<ParachainClient>, ParachainClient>,
```
Comment (skunert) [inline docs/sdk/src/guides/enable_elastic_scaling.rs:93]: I think we don't need `SlotBasedBlockImport` anymore if we only support 6s slots or more. wdyt @bkchr

**Review by skunert — COMMENTED** [review]
Overall looking pretty good!

**Comment by alindima** [comment]
Superseded by https://github.com/paritytech/polkadot-sdk/pull/9677
